### PR TITLE
Resolve issue #435

### PR DESCRIPTION
**Pull request by amzid**
This PR resolves the issue #435. The following steps were performed:

* **Implementation**:  
   * Separate done and failed payouts 
   * Add skipped status
   * Improve tzstats implementation for retry failed
@habanoz would be great if you have a look if the new changes are well-defined according to the state machine

* **Performed tests**: dry run in mainnet.

* **Documentation**: comments in code

**Work effort**: 3 hrs.

### CODE REVIEW

**Comment by jdsika** [comment]
btw, how can people override the decision that payments are skipped with "fee > payout amount" ?

File: src/Constants.py
Comment (habanoz) [inline src/Constants.py:61]: SKIP is a term used in calculations. I find this confusing, a different name may be used. That is being said, What is the distinction between skipped and fail?

File: src/pay/batch_payer.py
Comment (habanoz) [inline src/pay/batch_payer.py:432]: I would argue DONE is better suited for this case.

File: src/tzstats/tzstats_reward_provider_helper.py
Comment (habanoz) [inline src/tzstats/tzstats_reward_provider_helper.py:186]: I am unable to understand the reason behind this change.

**Review by habanoz — CHANGES_REQUESTED** [review]
@amzid Please see my comments. I will do another review according to your reply.

**Comment by habanoz** [comment]
I have made a review, waiting for a reply to make another review. I will check state machine diagrams.

Please see my comment on the linked issue.
https://github.com/tezos-reward-distributor-organization/tezos-reward-distributor/issues/435#issuecomment-841720725

**Comment by habanoz** [comment]
https://github.com/tezos-reward-distributor-organization/tezos-reward-distributor/issues/435#issue-889940928

I do not know if it is part of the request but, the reason for skipping is not added anywhere.

**Comment by habanoz** [comment]
Review effort is 1H.

File: src/tzstats/tzstats_reward_provider_helper.py
Comment (amzid) [reply src/tzstats/tzstats_reward_provider_helper.py:186]: This is to improve the implementation of tztats for retry_failed mode, mainly the way that current balances are updated is done in batches now instead of single queries

File: src/Constants.py
Comment (amzid) [reply src/Constants.py:61]: * PAID/DONE status is for payment items which are successfully applied and are (or are expected to be) part of the blockchain and can be queried using a block explorer. When a delegator has a payout with a status PAID that means the baker is able to provide a proof on the blockchain that the reward payment was done successfully.
* FAILED status is for payment items which failed at any stage of the payment pipeline (run_operation/forging/signing/pre_apply). The reasons for that depend on the individual case and bakers in general need support to solve that (support queries take time and effort).
* SKIPPED status is for payment items which are avoided with intention. In most cases due to the fact that costs of the transaction are higher than the reward amount. For many bakers, this transaction is not failed because it is avoided with intention and it is not DONE because the payout was not really performed by the baker, thus the idea of this new status.

File: src/pay/batch_payer.py
Comment (amzid) [reply src/pay/batch_payer.py:432]: SKIPPED status is for payment items which are avoided with intention. In most cases due to the fact that costs of the transaction are higher than the reward amount. For many bakers, this transaction is not failed because it is avoided with intention and it is not DONE because the payout was not really performed by the baker, thus the idea of this new status.

**Comment by vkresch** [comment]
@amzid please change the baking address of `tz1Z1tMai15JWUWeN2PKL9faXXVPMuWamzJj` (TezosLand) to `tz1g8vkmcde6sWKaG2NN9WKzCkDM6Rziq194` (StakeNow) in tezos-reward-distributor/tests/unit/test_BakingYamlConfParser.py because TezosLand is not a valid baker anymore thus the CI is failing. Maybe we should patch the baker validation function.

**Comment by jdsika** [comment]
@vkresch sorry, could you fix it as he seems not to be responding?

**Comment by amzid** [comment]
@vkresch @jdsika I have just changed the test baking address! Thanks for the review!

**Review by jdsika — DISMISSED** [review]

**Comment by habanoz** [comment]
@amzid 

> For many bakers, this transaction is not failed because it is avoided with intention and it is not DONE because the payout was not really performed by the baker

DONE is used for zero amounts. It is not necessary to make any payments but as a task, it is completed. DONE is also used in the DRY mode for the same interpretation. If payout was performed final status would be PAID. 

> We could change the new of SKIPPED to AVOIDED or SKIPPED_DUE_TO_HIGH_FEES if you think that the name is misleading?

I am OK with new names. You may choose one of them.

File: src/pay/batch_payer.py
Comment (habanoz) [reply src/pay/batch_payer.py:432]: I think the new status is better after seeing your comments.

Comment (habanoz) [inline src/pay/batch_payer.py:432]: Since the payment amount is 0, this I believe should be DONE.

**Comment by habanoz** [comment]
>  if you have a look if the new changes are well-defined according to the state machine

For state changes I have created https://github.com/tezos-reward-distributor-organization/tezos-reward-distributor/issues/443

**Comment by habanoz** [comment]
Total effort 1H +0.5 H = 1.5 H

**Comment by amzid** [comment]
Thank you for the feedback and explanations @habanoz. I also agree with the outcome. I will revert the done status for zero amounts and rename the new status.

**Review by habanoz — APPROVED** [review]

**Pull request merged**
